Reference/Common-Pitfalls/index-v9.md

```diff
@@ -0,0 +1,426 @@
+{
+    // The services are injected with DI and passed to the parent class
+    public ContactFormSurfaceController(
+        IUmbracoContextAccessor umbracoContextAccessor,
+        IUmbracoDatabaseFactory databaseFactory,
+        ServiceContext services,
+        AppCaches appCaches,
+        IProfilingLogger profilingLogger,
+        IPublishedUrlProvider publishedUrlProvider)
+        : base(umbracoContextAccessor, databaseFactory, services, appCaches, profilingLogger, publishedUrlProvider)
+    {
+    }
+
+    [HttpPost]
+    public IActionResult SubmitForm(ContactFormModel model)
+    {
+        // All normal form processing logic is left out of this example for brevity
+        // You can access all of these because they are properties of the base class, 
+        // if you need something else you can inject it in the constructor.
+        
+        //Profiling logger
+        using (ProfilingLogger.TraceDuration<ContactFormSurfaceController>("Start", "stop"))
+        {
+            // UmbracoContext
+            UmbracoContext.Content.GetById(1234);
+        }
+
+        return Ok();
+    }
+}
+```
+
+So next time you are using a singleton pattern or a static, think "Why am I doing this?", "I'm using Dependency Injection, I should be injecting this instance into my class."
+
+## Static references to scoped instances (such as `UmbracoHelper`)
+
+__Example 1:__
+
+```csharp
+public class BadApiController : UmbracoApiController
+{
+    // Note that this is static, that's bad
+    private static UmbracoHelper _umbracoHelper;
+    
+    public BadApiController(IUmbracoHelperAccessor umbracoHelperAccessor)
+    {
+        // Don't do this, this is bad
+        if (_umbracoHelper is null)
+        {
+            umbracoHelperAccessor.TryGetUmbracoHelper(out UmbracoHelper umbracoHelper);
+            _umbracoHelper = umbracoHelper;
+        }
+    }
+}
+```
+
+This practice can cause memory leaks along with inconsistent data results when using this `_umbracoHelper` instance.
+
+__Why?__
+
+It's important to understand the difference between an object that has a Request based scope/lifespan and
+an object that has an Singleton/Application based scope/lifespan ... here's the gist:
+
+* Application scope - if an object has an singleton/application scope/lifespan, that means that this single object
+instance will exist for the lifetime of the application. The single instance will be shared by every thread that
+accesses it. Static variables will always be application lifespan.
+* Request scope - The web world is made up of requests and each request has its own thread. When an object is in the scope of a Request it only survives as long as the web request survives. At the end of the web request, it may either be disposed or cleared from memory by the garbage collector. Request scoped object instances are not accessed by every other thread in the application - __unless you do something like the above!__
```

Comment:
```suggestion
__Why?__

It's important to understand the difference between an object that has a Request based scope/lifespan and
an object that has a Singleton/Application based scope/lifespan ... here's the gist:

* Application scope - if an object has a singleton/application scope/lifespan, that means that this single object
instance will exist for the lifetime of the application. The single instance will be shared by every thread that
accesses it. Static variables will always be application lifespan.
* Request scope - The web world is made up of requests and each request has its own thread. When an object is in the scope of a Request it only survives as long as the web request survives. At the end of the web request, it may either be disposed of or cleared from memory by the garbage collector. Request scoped object instances are not accessed by every other thread in the application - __unless you do something like the above!__
```